tomcat-gcp-lib/src/main/java/com/google/cloud/runtimes/tomcat/session/DatastoreStore.java

```diff
@@ -204,39 +233,79 @@ public void clear() throws IOException {
 
-    ByteArrayOutputStream bos = new ByteArrayOutputStream();
-
-    try (ObjectOutputStream oos = new ObjectOutputStream(bos)) {
-      if (!(session instanceof StandardSession)) {
-        throw new IOException(
-            "The session must be an instance of StandardSession to be serialized");
-      }
-      ((StandardSession) session).writeObjectData(oos);
-    }
-
-    Entity.Builder sessionEntity = Entity.newBuilder(keyFactory.newKey(session.getId()))
-        .set("content", Blob.copyFrom(bos.toByteArray()))
-        .set("lastAccess", session.getLastAccessedTime());
-
-    // A negative time indicates that the session should never time out
-    if (session.getMaxInactiveInterval() >= 0) {
-      sessionEntity.set("expirationTime",
-          session.getLastAccessedTime() + session.getMaxInactiveInterval() * 1000);
+    if (!(session instanceof DatastoreSession)) {
+      throw new IOException(
+          "The session must be an instance of DatastoreSession to be serialized");
     }
+    DatastoreSession datastoreSession = (DatastoreSession) session;
+    Key sessionKey = newKey(session.getId());
+    KeyFactory attributeKeyFactory = datastore.newKeyFactory()
+        .setKind(sessionKind)
+        .addAncestor(PathElement.of(sessionKind, sessionKey.getName()));
+
+    List<Entity> entities = serializeSession(datastoreSession, sessionKey, attributeKeyFactory);
+
+    TraceContext datastoreSaveContext = startSpan("Storing the session in the Datastore");
+    datastore.put(entities.toArray(new FullEntity[0]));
+    datastore.delete(datastoreSession.getSuppressedAttributes().stream()
+        .map(attributeKeyFactory::newKey)
+        .toArray(Key[]::new));
+    endSpan(datastoreSaveContext);
+  }
 
-    datastore.put(sessionEntity.build());
+  /**
+   * Serialize a session to a list of Entities that can be stored to the Datastore.
+   * @param session The session to serialize.
+   * @return A list of one or more entities containing the session and its attributes.
+   * @throws IOException If the session cannot be serialized.
+   */
+  @VisibleForTesting
+  List<Entity> serializeSession(DatastoreSession session, Key sessionKey,
+      KeyFactory attributeKeyFactory) throws IOException {
+    TraceContext serializationContext = startSpan("Serialization of the session");
+    List<Entity> entities = session.saveToEntities(sessionKey, attributeKeyFactory);
+    endSpan(serializationContext);
+    return entities;
   }
 
+  /**
+   * Remove expired sessions from the datastore.
+   */
   @Override
   public void processExpires() {
     log.debug("Processing expired sessions");
 
-    Query<Key> query = keyQueryBuilder
-        .setFilter(PropertyFilter.le("expirationTime", System.currentTimeMillis()))
+    Query<Key> query = Query.newKeyQueryBuilder().setKind(sessionKind)
+        .setFilter(PropertyFilter.le(SessionMetadata.EXPIRATION_TIME.getValue(),
+            System.currentTimeMillis()))
```

Comment:
extract this to a Clock interface for testability?

Reply:
Yes, so far the tests are mocking the Datastore response and therefore do not take into account the time, but I will refactor that to a Clock in case we want to add more precise tests later.